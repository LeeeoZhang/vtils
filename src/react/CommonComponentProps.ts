import React from 'react'

/**
 * 常用的组件属性。
 */
export interface CommonComponentProps<TRef = never> {
  key?: React.Key
  ref?: React.Ref<TRef>
  className?: string
  style?: React.CSSProperties
}
