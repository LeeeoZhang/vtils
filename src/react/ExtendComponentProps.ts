import React from 'react'
import { Merge } from '../types'

/**
 * 继承组件属性。
 *
 * @template TComponent 组件
 * @template TExclude 需去除的属性
 * @template TRef 新的 ref
 * @template TOverride 覆盖属性
 */
export type ExtendComponentProps<
  TComponent extends
    | React.ComponentType
    | keyof JSX.IntrinsicElements
    | React.JSXElementConstructor<any>,
  TExclude extends keyof React.ComponentProps<TComponent>,
  TRef extends any = never,
  TOverride extends Partial<
    Record<keyof React.ComponentProps<TComponent>, any>
  > = never
> = Merge<
  Merge<
    Omit<React.ComponentProps<TComponent>, TExclude>,
    [TOverride] extends [never] ? {} : TOverride
  >,
  [TRef] extends [never] ? {} : React.RefAttributes<TRef>
>
