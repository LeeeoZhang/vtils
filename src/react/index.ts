/**
 * React 工具库。基于 {@link https://github.com/streamich/react-use#readme | react-use}。
 *
 * @packageDocumentation
 */

/* istanbul ignore file */

export * from 'react-use'

// @index(['./**/*.ts', '!./**/*.{test,taro}.*', '!./{useToggle,createGlobalState,useTitle,useInterval,useSearchParam,useLocalStorage,useWindowSize,useHover}.*'], f => `export * from '${f.path}'`)
export * from './defineComponent'
export * from './ExtendComponentProps'
export * from './isVisibleValue'
export * from './renderComponent'
export * from './useClassName'
export * from './useControllableValue'
export * from './useLoadMore'
export * from './useReachBottom'
export * from './useScrollLoadMore'
export * from './useValidator'
// @endindex

// 与 react-use 同名的应手动指定导出的模块
export { useToggle, UseToggleResult } from './useToggle'
export {
  createGlobalState,
  CreateGlobalStateResult,
  CreateGlobalStateCustomResult,
  CreateGlobalStateResultResult,
  CreateGlobalStateState,
} from './createGlobalState'
export { useTitle } from './useTitle'
export { useInterval, UseIntervalResult } from './useInterval'
export { useSearchParam } from './useSearchParam'
export { useLocalStorage, UseLocalStorageResult } from './useLocalStorage'
export { useWindowSize } from './useWindowSize'
export { useHover, UseHoverOptions, UseHoverResult } from './useHover'
