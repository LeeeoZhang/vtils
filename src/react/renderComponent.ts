import React from 'react'
import ReactDOM from 'react-dom'
import { PickBy } from 'vtils/types'

export interface RenderComponentResult<
  TComponent extends React.ComponentType<any>
> {
  /**
   * 增量重渲染，新属性将与增量属性合并并成为新的增量属性。
   *
   * @param props 新属性
   */
  incrementalRerender(props: Partial<React.ComponentProps<TComponent>>): void

  /**
   * 部分重渲染，新属性将与初始属性合并。
   *
   * @param props 新属性
   */
  partialRerender(props: Partial<React.ComponentProps<TComponent>>): void

  /**
   * 完全重渲染，新属性将直接作为全部属性传给组件。
   *
   * @param props 新属性
   */
  fullRerender(props: React.ComponentProps<TComponent>): void

  /**
   * 销毁组件并移除 DOM 节点。
   */
  destroy(): void
}

/**
 * 独立渲染一个组件在 document.body 下，常应用于弹窗类组件。
 *
 * @param Component 要渲染的组件
 * @param initialProps 初始属性
 * @param injectCallbacks 回调函数注入
 */
export function renderComponent<TComponent extends React.ComponentType<any>>(
  Component: TComponent,
  initialProps: React.ComponentProps<TComponent>,
  injectCallbacks?: PickBy<
    React.ComponentProps<TComponent>,
    Function | undefined
  >,
): RenderComponentResult<TComponent> {
  let container: HTMLDivElement | null = document.createElement('div')
  document.body.appendChild(container)

  const render = (props: Partial<React.ComponentProps<TComponent>>) => {
    props = { ...props }
    if (injectCallbacks) {
      for (const key of Object.keys(injectCallbacks)) {
        const originalCallback = props[key]
        ;(props as any)[key] = async () => {
          await originalCallback?.()
          await (injectCallbacks as any)[key]()
        }
      }
    }
    ReactDOM.render(React.createElement(Component, props), container)
  }

  render(initialProps)

  let incrementalProps: React.ComponentProps<TComponent> = { ...initialProps }

  return {
    incrementalRerender(props: Partial<React.ComponentProps<TComponent>>) {
      if (!container) return
      incrementalProps = {
        ...incrementalProps,
        ...props,
      }
      render(incrementalProps)
    },
    partialRerender(props: Partial<React.ComponentProps<TComponent>>) {
      if (!container) return
      render({
        ...initialProps,
        ...props,
      })
    },
    fullRerender(props: React.ComponentProps<TComponent>) {
      if (!container) return
      render(props)
    },
    destroy() {
      if (!container) return
      const unmountResult = ReactDOM.unmountComponentAtNode(container)
      if (unmountResult) {
        container.parentNode?.removeChild(container)
      }
      container = null
    },
  }
}
