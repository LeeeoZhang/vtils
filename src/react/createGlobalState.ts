import { Dispatch, SetStateAction, useEffect, useState } from 'react'
import { EventBus } from '../utils'

export interface CreateGlobalStateResult<S> {
  (): readonly [S, Dispatch<SetStateAction<S>>]
  getState(): S
  setState(nextState: SetStateAction<S>): void
  watchState(callback: (nextState: S, prevState: S) => any): () => void
}

export function createGlobalState<S>(): CreateGlobalStateResult<S | undefined>

export function createGlobalState<S>(
  initialState: S,
): CreateGlobalStateResult<S>

export function createGlobalState<S>(
  initialState?: S,
): CreateGlobalStateResult<S | undefined> {
  const bus = new EventBus<{
    setGlobalState: (
      nextGlobalState: S | undefined,
      prevGlobalState: S | undefined,
    ) => void
  }>()
  let currentGlobalState: S | undefined = initialState
  const getGlobalState: () => S | undefined = () => currentGlobalState
  const setGlobalState: Dispatch<SetStateAction<
    S | undefined
  >> = nextGlobalState => {
    if (typeof nextGlobalState === 'function') {
      nextGlobalState = (nextGlobalState as any)(currentGlobalState)
    }
    if (nextGlobalState !== currentGlobalState) {
      const prevGlobalState = currentGlobalState
      currentGlobalState = nextGlobalState as any
      bus.emit('setGlobalState', nextGlobalState as any, prevGlobalState)
    }
  }
  const watchGlobalState: (
    callback: (
      nextGlobalState: S | undefined,
      prevGlobalState: S | undefined,
    ) => any,
  ) => () => void = callback => {
    return bus.on('setGlobalState', callback)
  }
  const useGlobalState: CreateGlobalStateResult<S | undefined> = (() => {
    const [state, setState] = useState(currentGlobalState)
    useEffect(() => watchGlobalState(setState), [])
    return [state, setGlobalState] as const
  }) as any
  useGlobalState.getState = getGlobalState
  useGlobalState.setState = setGlobalState
  useGlobalState.watchState = watchGlobalState
  return useGlobalState
}
