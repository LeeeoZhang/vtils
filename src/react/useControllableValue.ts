import { Defined } from '../types'
import { useCallback, useState } from 'react'
import { useUpdateEffect } from 'react-use'

export type UseControllableValueOptions<
  TProps,
  TDefaultValuePropName extends keyof TProps,
  TValuePropName extends keyof TProps,
  TCallbackPropName extends keyof TProps,
  TDefaultValue extends TProps[TValuePropName]
> = {
  /**
   * 默认值的属性名。
   */
  defaultValuePropName: TDefaultValuePropName

  /**
   * 值的属性名。
   */
  valuePropName: TValuePropName

  /**
   * 值改变时的回调函数的属性名。
   */
  callbackPropName: TCallbackPropName

  /**
   * 默认值。
   */
  defaultValue?: TDefaultValue

  /**
   * 是否总是更新值。
   */
  alwaysUpdateValue?: boolean
}

export type UseControllableValueResult<
  TProps,
  TValuePropName extends keyof TProps,
  TCallbackPropName extends keyof TProps,
  TDefaultValue extends TProps[TValuePropName]
> = [
  TDefaultValue extends undefined
    ? TProps[TValuePropName]
    : Defined<TProps[TValuePropName]>,
  Defined<TProps[TCallbackPropName]>,
]

/**
 * 受控值。
 *
 * @param props 组件的属性
 * @param options 选项
 */
export function useControllableValue<
  TProps,
  TDefaultValuePropName extends keyof TProps,
  TValuePropName extends keyof TProps,
  TCallbackPropName extends keyof TProps,
  TDefaultValue extends TProps[TValuePropName]
>(
  props: TProps,
  options: UseControllableValueOptions<
    TProps,
    TDefaultValuePropName,
    TValuePropName,
    TCallbackPropName,
    TDefaultValue
  >,
): UseControllableValueResult<
  TProps,
  TValuePropName,
  TCallbackPropName,
  TDefaultValue
> {
  const [value, setValue] = useState(() => {
    if (options.valuePropName in props) {
      return props[options.valuePropName]
    }
    if (options.defaultValuePropName in props) {
      return props[options.defaultValuePropName]
    }
    return options.defaultValue
  })

  useUpdateEffect(() => {
    if (options.valuePropName in props) {
      setValue(props[options.valuePropName])
    }
  }, [props[options.valuePropName]])

  const handleSetValue = useCallback(
    (nextValue: typeof value) => {
      if (!(options.valuePropName in props) || options.alwaysUpdateValue) {
        setValue(nextValue)
      }
      if (typeof props[options.callbackPropName] === 'function') {
        ;(props[options.callbackPropName] as any)(nextValue)
      }
    },
    [props, options.valuePropName, options.callbackPropName],
  )

  return [value, handleSetValue] as any
}
