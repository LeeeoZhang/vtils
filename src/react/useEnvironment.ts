import { getEnvironment, GetEnvironmentResult } from '../utils'
import { useMemo } from 'react'

/**
 * 获取运行环境信息。
 *
 * @returns 返回运行环境信息
 */
export function useEnvironment(): GetEnvironmentResult {
  return useMemo(getEnvironment, [])
}
