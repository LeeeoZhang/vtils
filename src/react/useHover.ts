import { useCallback, useRef, useState } from 'react'
import { useLatest, useUnmount } from 'react-use'

export interface UseHoverOptions {
  /**
   * 悬停开始后多久开启悬停态，单位为毫秒。
   *
   * @default 50
   */
  hoverStartDelay?: number

  /**
   * 悬停结束后多久关闭悬停态，单位为毫秒。
   *
   * @default 400
   */
  hoverEndDelay?: number
}

export interface UseHoverResult {
  /**
   * 当前的悬停态。
   */
  hovering: boolean

  /**
   * 开始悬停。
   */
  startHover: () => any

  /**
   * 结束悬停。
   */
  endHover: () => any
}

export function useHover(options: UseHoverOptions = {}): UseHoverResult {
  const { hoverStartDelay = 50, hoverEndDelay = 400 } = options

  const [hovering, setHovering] = useState(false)

  const latestHoverStartDelay = useLatest(hoverStartDelay)
  const latestHoverEndDelay = useLatest(hoverEndDelay)
  const latestHovering = useLatest(hovering)

  const hoverStartTimer = useRef<any>()
  const hoverEndTimer = useRef<any>()

  const clearHoverStartTimer = useCallback(() => {
    if (hoverStartTimer.current) {
      clearTimeout(hoverStartTimer.current)
      hoverStartTimer.current = undefined
    }
  }, [])
  const clearHoverEndTimer = useCallback(() => {
    if (hoverEndTimer.current) {
      clearTimeout(hoverEndTimer.current)
      hoverEndTimer.current = undefined
    }
  }, [])

  const startHover = useCallback(() => {
    if (!latestHovering.current) {
      clearHoverStartTimer()
      hoverStartTimer.current = setTimeout(() => {
        setHovering(true)
      }, latestHoverStartDelay.current)
    }
  }, [])
  const endHover = useCallback(() => {
    if (latestHovering.current) {
      clearHoverEndTimer()
      hoverEndTimer.current = setTimeout(() => {
        setHovering(false)
      }, latestHoverEndDelay.current)
    } else {
      clearHoverStartTimer()
    }
  }, [])

  useUnmount(() => {
    clearHoverStartTimer()
    clearHoverEndTimer()
  })

  return {
    hovering,
    startHover,
    endHover,
  }
}
