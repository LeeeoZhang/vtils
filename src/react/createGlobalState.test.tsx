import React from 'react'
import { act, renderHook } from '@testing-library/react-hooks'
import { createGlobalState, CreateGlobalStateResult } from './createGlobalState'
import { render } from '@testing-library/react'

describe('createGlobalState', () => {
  describe('有初始值', () => {
    let useGlobalName: CreateGlobalStateResult<string>

    beforeEach(() => {
      useGlobalName = createGlobalState('')
    })

    test('getState 正常', () => {
      expect(useGlobalName.getState()).toBe('')
    })

    test('setState 正常', () => {
      useGlobalName.setState('Jay')
      expect(useGlobalName.getState()).toBe('Jay')
    })

    test('watchState 正常', () => {
      const cb = jest.fn()
      const off = useGlobalName.watchState(cb)
      useGlobalName.setState('Jay')
      expect(cb).toBeCalled().toBeCalledTimes(1).toBeCalledWith('Jay', '')
      useGlobalName.setState('Jay2')
      expect(cb).toBeCalled().toBeCalledTimes(2).toBeCalledWith('Jay2', 'Jay')
      off()
      expect(cb).toBeCalled().toBeCalledTimes(2).toBeCalledWith('Jay2', 'Jay')
    })

    test('跨组件状态共享正常', () => {
      const { result: result1 } = renderHook(() => useGlobalName())
      const { result: result2 } = renderHook(() => useGlobalName())
      const { result: result3 } = renderHook(() => useGlobalName())

      expect(useGlobalName.getState())
        .toBe(result1.current[0])
        .toBe(result2.current[0])
        .toBe(result3.current[0])

      act(() => result2.current[1]('Fong'))

      expect(useGlobalName.getState())
        .toBe(result1.current[0])
        .toBe(result2.current[0])
        .toBe(result3.current[0])
        .toBe('Fong')

      act(() => result1.current[1](() => 'jj'))

      expect(useGlobalName.getState())
        .toBe(result1.current[0])
        .toBe(result2.current[0])
        .toBe(result3.current[0])
        .toBe('jj')
    })
  })

  describe('无初始值', () => {
    let useGlobalName: CreateGlobalStateResult<string | undefined>

    beforeEach(() => {
      useGlobalName = createGlobalState()
    })

    test('getState 正常', () => {
      expect(useGlobalName.getState()).toBe(undefined)
    })

    test('setState 正常', () => {
      useGlobalName.setState('Jay')
      expect(useGlobalName.getState()).toBe('Jay')
    })

    test('watchState 正常', () => {
      const cb = jest.fn()
      const off = useGlobalName.watchState(cb)
      useGlobalName.setState('Jay')
      expect(cb)
        .toBeCalled()
        .toBeCalledTimes(1)
        .toBeCalledWith('Jay', undefined)
      useGlobalName.setState('Jay2')
      expect(cb).toBeCalled().toBeCalledTimes(2).toBeCalledWith('Jay2', 'Jay')
      off()
      expect(cb).toBeCalled().toBeCalledTimes(2).toBeCalledWith('Jay2', 'Jay')
    })

    test('跨组件状态共享正常', () => {
      const { result: result1 } = renderHook(() => useGlobalName())
      const { result: result2 } = renderHook(() => useGlobalName())
      const { result: result3 } = renderHook(() => useGlobalName())

      expect(useGlobalName.getState())
        .toBe(result1.current[0])
        .toBe(result2.current[0])
        .toBe(result3.current[0])

      act(() => result2.current[1]('Fong'))

      expect(useGlobalName.getState())
        .toBe(result1.current[0])
        .toBe(result2.current[0])
        .toBe(result3.current[0])
        .toBe('Fong')

      act(() => result1.current[1](() => 'jj'))

      expect(useGlobalName.getState())
        .toBe(result1.current[0])
        .toBe(result2.current[0])
        .toBe(result3.current[0])
        .toBe('jj')
    })
  })

  describe('bug', () => {
    test('延后注册的订阅不能获取最新值', async () => {
      const useGlobalName = createGlobalState('o')

      function Child() {
        const [name] = useGlobalName()
        return <div id='child'>{name}</div>
      }

      function Parent() {
        const [name] = useGlobalName()
        return (
          <>
            <div id='parent'>{name}</div>
            {name === 'ok' && <Child />}
          </>
        )
      }

      const { asFragment } = render(<Parent />)
      expect(asFragment()).toMatchSnapshot('只有 parent 且值为 o')
      useGlobalName.setState('ok')
      expect(asFragment()).toMatchSnapshot('有 parent 和 child 且值都为 ok')
    })
  })
})
