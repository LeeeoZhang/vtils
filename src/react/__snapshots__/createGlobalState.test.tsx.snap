// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createGlobalState bug 延后注册的订阅不能获取最新值: 只有 parent 且值为 o 1`] = `
<DocumentFragment>
  <div
    id="parent"
  >
    o
  </div>
</DocumentFragment>
`;

exports[`createGlobalState bug 延后注册的订阅不能获取最新值: 有 parent 和 child 且值都为 ok 1`] = `
<DocumentFragment>
  <div
    id="parent"
  >
    ok
  </div>
  <div
    id="child"
  >
    ok
  </div>
</DocumentFragment>
`;
